lib/puppet/type/file/content.rb

```diff
@@ -119,6 +120,20 @@ def insync?(is)
       result
     end
 
+    def property_matches?(current, desired)
+      basic = super
+      # The inherited equality is always accepted, so use it if valid.
+      time_types = ["mtime", "ctime"]
+      return basic if basic || !time_types.include?(checksum_type.to_s)
```

Comment:
It turns out that the `#checksum_type` is essentially dead code. It's only called here and in `content_spec`, and I think can be deleted. I think it's existence is due to the fact that the checksum type was not always set correctly due to all of the issues you've fixed.

If the file resource has a `source`, then the `checksum_type` and `checksum` are copied from the metadata to the `content` and `checksum` parameter values, and become the `desired` value:

```
(rdb:1) current
"{mtime}2015-02-03 23:34:26 -0800"
(rdb:1) desired
"{mtime}2015-02-03 21:58:49 -0800"
(rdb:1) resource.parameter(:content).value
"{mtime}2015-02-03 21:58:49 -0800"
(rdb:1) resource.parameter(:source).checksum
"{mtime}2015-02-03 21:58:49 -0800"
```

And the `checksum` parameter already contains what the `checksum_type` method was trying to do:

```
(rdb:1) resource.parameter(:checksum).value
:mtime
```

If the file resource has `content`, then the checksum is stored as the value of the `content` parameter, and it becomes the `desired` value:

```
(rdb:1) current
"{sha256}ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad"
(rdb:1) desired
"{sha256}88d4266fd4e6338d13b845fcf289579d209c897823b9217da3e161936f031589"
(rdb:1) resource.parameter(:content).value
"{sha256}88d4266fd4e6338d13b845fcf289579d209c897823b9217da3e161936f031589"
(rdb:1) resource.parameter(:source)
nil
```

And again the `checksum` parameter contains the expected value:

```
(rdb:1) resource.parameter(:checksum).value
:sha256
```

So I think we can delete `checksum_type`, and just call `resource.parameter(:checksum).value` directly.